🎁 Add "Surround with snippet" to quick fix menu

babakks opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR fixes #150678

Summary of changes:

- Thc core functionality of "Surround with snippet" is extracted into a new class, `SurroundWithSnippet`.
- A class extending `EditorAction2` is added to expose the command.
- A class implementing `CodeActionProvider` is added to expose the command as a code action (quick fix).

babakks commented:
@jrieken The workflow failure seems to be unrelated to the changes here.

jrieken commented on src/vs/workbench/contrib/snippets/browser/surroundWithSnippet.ts:
```diff
@@ -4,57 +4,136 @@
+
+export class SurroundWithSnippetCodeActionProvider extends Disposable implements CodeActionProvider {
+	private static readonly codeAction: CodeAction = {
+		kind: CodeActionKind.QuickFix.value,
```
The `Refactor`-type should be used here

babakks replied:
Done.

jrieken commented on src/vs/workbench/contrib/snippets/browser/surroundWithSnippet.ts:
```diff
@@ -4,57 +4,136 @@
+		}
+		const snippets = await this.core.getSurroundableSnippets();
+		return {
+			actions: snippets.length ? [SurroundWithSnippetCodeActionProvider.codeAction] : [],
```
This will just invoke the picker but I think it would be better to directly show the snippets. Only show the picker when there is too much to pick from (>7 or so). 

For each surroundable snippet you can use [InsertSnippetAction](https://github.com/microsoft/vscode/blob/1c5723822a33d7c40913f98d8a2f639b40e7f8ad/src/vs/workbench/contrib/snippets/browser/insertSnippet.ts#L48) and bind the snippet-argument accordingly

jrieken commented on src/vs/workbench/contrib/snippets/browser/surroundWithSnippet.ts:
```diff
@@ -4,57 +4,136 @@
+		}
 
-	async runEditorCommand(accessor: ServicesAccessor, editor: ICodeEditor, ...args: any[]) {
+		const model = this._editor.getModel();
```
Only extract this into a seprate function, e.g `function surroundateSnippet(accessor, model)` and leave the action as-is otherwise

babakks replied:
I didn't fully understand this commment. So, I took this as a hint to refactor from class-based to function-based implementation. So, I extracted the functions out and  added an `accessor: ServicesAccessor` parameter to them to help resolve their dependencies.

jrieken reviewed (changes requested):
Thanks so far. This is already good but I think we can make this more smooth. Less click ftw!

jrieken commented on src/vs/workbench/contrib/snippets/browser/surroundWithSnippet.ts:
```diff
@@ -4,57 +4,136 @@
+	}
+}
+
+registerEditorContribution(options.id, SurroundWithSnippetCodeActionProvider);
```
Oh, forget to mention that we need something better than an editor contribution. Those get created per editor but `languageFeaturesService.codeActionProvider.register` is gobal

jrieken commented:
Another note: fresh from the press we are working on supporting snippets with workspace edits (https://github.com/microsoft/vscode/issues/145374). So, this feature could be build even better, e.g don't have commands but use a workspace edit (for each surround-with snippet) that uses the new snippet edit support.

babakks commented:
@jrieken After changing from editor contribution to workbench contribution, strangely, context key data were not available to check the command precondition at `canExecute` function. So, I opted out of context-based checking and went toward implementing `canExecute` by using `IEditorService`.

You can check it by replacing the `canExecute` function with this:

```ts
function canExecute(accessor: ServicesAccessor): boolean {
    return accessor.get(IContextKeyService).contextMatchesRules(options.precondition);
}
```

jrieken commented:
I was thinking different and that this can be done simpler. The code actions should not use the `command` property but the `edit` property. That workspace edit should have a single `IWorkspaceTextEdit`-edit which uses the new [`insertAsSnippet`](https://github.com/microsoft/vscode/blob/f413297170178f16ab218202153b629d59a98be1/src/vs/editor/common/languages.ts#L1424)-flag. That's it. Than have a single "overflow"-command when there are too many snippets which simply invokes `editor.action.surroundWithSnippet` (e.g that one code action uses the `command` property)

* only extract the `getSurrounableSnippet` function
* leave `SurroundWithAction` unchanged otherwise
* have a code action provider that uses `getSurrounableSnippet` and that uses workspace edits for its actions

babakks commented:
@jrieken Thanks for the precise details. I've made the changes.

However, there was a bug, I think, with the `SnippetSession.createEditsAndSnippetsFromEdits` wherein variables were resolved after the replacement texts were rendered. So, I changed the order of statements and also added a test to verify this is now working.

Could you please review the changes?

babakks commented:
@jrieken Can't we also change the overflow command to use the new `WorkspaceTextEdit` approach? I mean, are we deliberately not changing the old behavior?

jrieken commented:
@babakks There are some things that I want to tweak before merging - mostly code style/shape. It looks like you need to enable this for this PR.

babakks commented:
The "Allow edits..." is enabled. Is there any other permission?

babakks commented:
@jrieken I just merged this with your latest changes.

jrieken commented:
> The "Allow edits..." is enabled. Also, I now added your account to my fork collaborators.

It finally worked - ran into some permission issues before 🤷. Excited to see this land

bpasero commented:
Awesome 👏 !

babakks commented:
@jrieken You're welcome. Thank you.
